src/iceberg/update/update_partition_spec.cc

```diff
@@ -0,0 +1,488 @@
+      } else {
+        new_fields.push_back(field);
+      }
+    } else if (format_version_ < 2) {
+      // In V1, deleted fields are replaced with void transform
+      auto rename_it = renames_.find(std::string(field.name()));
+      std::string field_name =
+          rename_it != renames_.end() ? rename_it->second : std::string(field.name());
+      new_fields.emplace_back(field.source_id(), field.field_id(), field_name,
+                              Transform::Void());
+    }
+    // In V2, deleted fields are simply removed
+  }
+
+  // Add new fields
+  new_fields.insert(new_fields.end(), adds_.begin(), adds_.end());
+
+  // Determine the new spec ID
+  int32_t new_spec_id = spec_ ? spec_->spec_id() + 1 : PartitionSpec::kInitialSpecId;
+
+  // In V2, if all fields are removed, reset last_assigned_partition_id to allow
+  // field IDs to restart from 1000 when fields are added again
+  int32_t last_assigned_id = last_assigned_partition_id_;
+  if (format_version_ >= 2 && new_fields.empty()) {
+    last_assigned_id = PartitionSpec::kLegacyPartitionDataIdStart - 1;
+  }
+
+  ICEBERG_ASSIGN_OR_RAISE(
+      auto new_spec,
+      PartitionSpec::Make(*schema_, new_spec_id, std::move(new_fields),
+                          /*allow_missing_fields=*/false, last_assigned_id));
+
+  ApplyResult result;
+  auto spec_id = new_spec->spec_id();
+
+  // Add the new partition spec
+  result.updates.emplace_back(
+      std::make_unique<table::AddPartitionSpec>(std::move(new_spec)));
+
+  // If set_as_default_ is true, set this spec as the default
+  if (set_as_default_) {
+    result.updates.emplace_back(
+        std::make_unique<table::SetDefaultPartitionSpec>(spec_id));
+  }
+
+  return result;
+}
+
+int32_t UpdatePartitionSpec::AssignFieldId() { return ++last_assigned_partition_id_; }
+
+PartitionField UpdatePartitionSpec::RecycleOrCreatePartitionField(
+    int32_t source_id, std::shared_ptr<Transform> transform, const std::string* name) {
+  // In V2+, use pre-built index for O(1) lookup instead of O(n*m) iteration
+  if (format_version_ >= 2 && !historical_fields_.empty()) {
+    const std::string transform_str = transform->ToString();
+    TransformKey key{source_id, transform_str};
+    auto it = historical_fields_.find(key);
+    if (it != historical_fields_.end()) {
+      const auto& field = it->second;
+      // If target name is specified then consider it too, otherwise not
+      if (name == nullptr || std::string(field.name()) == *name) {
+        return field;
+      }
+    }
+  }
+  // No matching field found, create a new one
+  std::string field_name = name ? *name : "";
+  return {source_id, AssignFieldId(), field_name, transform};
+}
+
+std::string UpdatePartitionSpec::GeneratePartitionName(
+    int32_t source_id, const std::shared_ptr<Transform>& transform) const {
+  // Find the source field name
+  auto field_result = schema_->FindFieldById(source_id);
+  std::string_view source_name = "unknown";
```

Comment:
What about returning Result and handling return for `FindFieldById` and unknown name?